src/glang/graphene_parser.py

```diff
@@ -582,8 +584,20 @@ def _function_call_outer(
         # resolution figure it out, although this isn't very explicit.
         if isinstance(node, lp.UFCS_Call):
             this_arg = self.parse_expr(node.expression)
-            if this_arg.expression().is_indirect_pointer_to_type:
+            # TODO: should there be syntax for this? It seems a bit adhoc
+            indirection_kind = this_arg.expression().underlying_indirection_kind
+            if indirection_kind == cg.Type.Kind.CONST_REF:
                 this_arg.add_parent(cg.BorrowExpression(this_arg.expression(), False))
+            elif indirection_kind == cg.Type.Kind.MUTABLE_REF:
+                this_arg.add_parent(cg.BorrowExpression(this_arg.expression(), True))
+            elif indirection_kind == cg.Type.Kind.VALUE:
```

Comment:
The `Type` we give to `TypedExpression` is exactly what the user would see if they ran something like `typeof(expression)`. `underlying_indirection_kind` records what the codegen's actually doing under the hood (which is different) eg. variable references are 
always passed around stack pointers not value types.

You can think of `underlying_indirection_kind` as the answer to "What is the most privileged way I can borrow this expression?". When `expression.underlying_type` is a type (which can only occur after a borrow), `underlying_indirection_kind` is always `Type.Kind.VALUE` since we've already borrowed the expression and we don't allow double indirections (so we cant borrow it any more than we already have).